### PR TITLE
DIG import

### PR DESCRIPTION
**Pull request by mattboehm**
Add an "Add dataset from DIG File" button to organization pages.

When clicked, this presents a form for the user to upload a DIG excel file. Upon successful upload, a new dataset is created with metadata extracted from the relevant fields.

CFPB employees can see a walkthrough of the functionality at https://cfpb.box.com/s/snssehnj9prfsgsz0io7jxgcef075e3o

### CODE REVIEW

**Comment by mattboehm** [comment]
I have not been able to test this against an actual DIG file yet as I can only run CKAN on my contractor-issued laptop and did not want to put any DIG files on it without receiving explicit permission.

Testing against actual files may reveal bugs that require future fixes, but I have tested against mock DIG files to verify the functionality.

**Comment by mattboehm** [comment]
Travis is failing due to `openpyxl` not being importable, but it is listed as a dependency in `setup.py`. Is there something I need to tweak, or do the Travis build steps need to somehow be modified?

**Comment by JeffreyMFarley** [comment]
`openpyxl` is what I used to read an Excel spreadsheet from within python.  It may be that this package needs approval?

File: ckanext/cfpb_extrafields/controllers/tst_import.py
Comment (JeffreyMFarley) [inline ckanext/cfpb_extrafields/controllers/tst_import.py]: We should think about how to provide a test file that we can check in

File: ckanext/cfpb_extrafields/digutils.py
Comment (JeffreyMFarley) [inline ckanext/cfpb_extrafields/digutils.py:1]: I am open to other suggestions

**Review by JeffreyMFarley — APPROVED** [review]
It's nice to see pieces of my prototype are going somewhere!

**Comment by JeffreyMFarley** [comment]
Let's wait on the merge until we see what to do with `openpyxl`

File: ckanext/cfpb_extrafields/controllers/tst_import.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/controllers/tst_import.py]: Yes, I should probably remove this entire file for the time being. There's unit tests for the pure function, but if there's a dig file that folks feel comfortable putting on the public github, I could make unit tests around that.

**Comment by mattboehm** [comment]
@JeffreyMFarley I know there are a few choices for 3rd-party libraries that work with xlsx formats, but my guess is that any particular one would need approval. Is there a list of approved libraries somewhere, or documentation on how to get additional packages approved?

**Comment by JeffreyMFarley** [comment]
> Travis is failing due to openpyxl not being importable, but it is listed as a dependency in setup.py. Is there something I need to tweak, or do the Travis build steps need to somehow be modified 

Your `travis.yml` runs `pip install -r requirements.txt` not `setup.py`.  Just make the change there and it should work.

Will get back to you about internal approvals, or Tim may be of assistance

**Comment by mattboehm** [comment]
@tanderegg Since this PR adds a dependency on openpyxl, is there any process that needs to happen before this is merged in?

@JeffreyMFarley I got the Travis tests to work by adding openpyxl to requirements-test and mocking a CKAN exception in one additional place.

**Comment by mattboehm** [comment]
By the way, we should tell coveralls to report coverage for `digutils.py` and `exportutils.py` (both are in the same directory as `validators.py`). Not sure who has admin access to coveralls; maybe @tanderegg or @JeffreyMFarley ?

At this point the merge is just awaiting some sort of thumbs-up saying that it's okay to depend on `openpyxl`. I asked someone to look into this (Orlando, I believe) at today's release management meeting, so hopefully that can happen soon.

**Comment by tanderegg** [comment]
@mattboehm For coveralls, you'll have to modify the .travis.yml file here: https://github.com/cfpb/ckanext-cfpb-extrafields/blob/master/.travis.yml#L8, right now it just points to "validators"

File: ckanext/cfpb_extrafields/controllers/digimport.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/controllers/digimport.py]: @mattboehm It might be good to add `as e` to the except statement, and include the error message in the string stored in `errors` to aid with future troubleshooting.

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/digutils.py]: It what scenario does the import fail, out of curiosity?

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/digutils.py]: Can we format this comment like the others with proper spacing?  Picky I know, but I like clean comments :)

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/digutils.py]: Add spacing to comment, also does this mean that the contents of B54 is used for "dataset_notes", "transfer_details", "website_url", and "wiki_link"?

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/digutils.py]: I think `callable(cell_or_func)` is a bit more exact and easy to read

File: ckanext/cfpb_extrafields/plugin.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/plugin.py]: Should we resolve this TODO before merging?

File: ckanext/cfpb_extrafields/tests.py
Comment (tanderegg) [inline ckanext/cfpb_extrafields/tests.py]: While we're at it, might as well remove the extra line

File: requirements-test.txt
Comment (tanderegg) [inline requirements-test.txt:6]: Do we need to add `lxml` here as well?

**Review by tanderegg — COMMENTED** [review]
Great work @mattboehm, excellent tests by the way.  Just a few comments/suggestions, thanks!

File: ckanext/cfpb_extrafields/digutils.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/digutils.py]: When we run the unit tests on travis, we don't install CKAN. The files that are called during unit tests (validators, digutils, exportutils) contain as much of the testable logic as possible without depending on CKAN to run.

File: ckanext/cfpb_extrafields/digutils.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/digutils.py]: I had copied these fields from the original code so I'm not positive, but I believe the idea was that B54 _could_ contain any of those pieces of information. I'm going to talk this over with Angie and either leave 3 of those fields blank or have them add separate fields for them if they want them ingested.

File: ckanext/cfpb_extrafields/digutils.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/digutils.py]: agreed; I'll change this.

File: ckanext/cfpb_extrafields/plugin.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/plugin.py]: yeah, I'll delete that.

File: requirements-test.txt
Comment (mattboehm) [reply requirements-test.txt:6]: We may as well. None of the current unit tests actually call openpyxl, but it needed to be included as it's still imported from digutils.

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [reply ckanext/cfpb_extrafields/digutils.py]: Ah, that makes sense, thanks!

File: ckanext/cfpb_extrafields/digutils.py
Comment (tanderegg) [reply ckanext/cfpb_extrafields/digutils.py]: Cool, sounds good, otherwise I feel like it could be confusing for folks.

File: requirements-test.txt
Comment (tanderegg) [reply requirements-test.txt:6]: Great

File: ckanext/cfpb_extrafields/controllers/digimport.py
Comment (mattboehm) [reply ckanext/cfpb_extrafields/controllers/digimport.py]: This block is only reached on a JSONDecodeError, which would probably only happen if a user tried to manually visit a url like `/import?errors=not_valid_json`. I can change the error message to be something more descriptive perhaps, like "The 'errors' parameter is not valid JSON". This would probably be more accurate than the exception's message (something like "Invalid character at position 54: ,"). Thoughts?

Comment (tanderegg) [reply ckanext/cfpb_extrafields/controllers/digimport.py]: Perhaps do both?  I guess this is a very unlikely scenario, but including the info on where the JSON decode error could be useful if we actually need to debug something.

**Review by tanderegg — APPROVED** [review]

**Pull request merged**
